tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart

```diff
@@ -1407,34 +1420,96 @@ class Restrictions {
         .classDefinition;
   }
 
-  ClassDefinition? _findTableClassInParentClasses(
+  /// Traverses up the class hierarchy from [currentModel], applying [findFunction] to each parent.
+  /// Returns the first non-null result from [findFunction], or `null` if no match is found.
+  ///
+  /// ```dart
+  /// var serverOnlyAncestor = _findInHierarchy(
+  ///   currentModel,
+  ///  (ClassDefinition ancestor) => ancestor.serverOnly ? ancestor : null,
+  /// );
+  /// ```
+  T? _findInHierarchy<T>(
     SerializableModelDefinition? currentModel,
+    T? Function(ClassDefinition) findFunction,
   ) {
     if (currentModel is! ClassDefinition) return null;
     var parentModel = _getParentClass(currentModel);
 
     while (parentModel != null) {
-      if (_hasTableDefined(parentModel)) {
-        return parentModel;
-      }
+      var result = findFunction(parentModel);
+      if (result != null) return result;
 
       parentModel = _getParentClass(parentModel);
     }
 
     return null;
   }
 
+  ClassDefinition? _findTableClassInParentClasses(
+    SerializableModelDefinition? currentModel,
+  ) {
+    return _findInHierarchy(
+      currentModel,
+      (ClassDefinition ancestor) =>
+          ancestor.tableName != null ? ancestor : null,
+    );
+  }
+
   ClassDefinition? _findServerOnlyClassInParentClasses(
     SerializableModelDefinition? currentModel,
   ) {
-    if (currentModel is! ClassDefinition) return null;
-    var parentModel = _getParentClass(currentModel);
+    return _findInHierarchy(
+      currentModel,
+      (ClassDefinition ancestor) => ancestor.serverOnly ? ancestor : null,
+    );
+  }
 
-    while (parentModel != null) {
-      if (parentModel.serverOnly) return parentModel;
+  ClassDefinition? _findAncestorWithDuplicatedFieldName(
+    SerializableModelDefinition? currentModel,
+  ) {
+    return _findInHierarchy(
+      currentModel,
+      (ClassDefinition ancestor) {
```

Comment:
Suggestion: You validate that it is a class definition on the next line. This means that you should probably take in a `SerializableModelDefinition` in the callback.